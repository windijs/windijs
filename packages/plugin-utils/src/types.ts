import type { Config } from "@windijs/helpers";

export type Frameworks = "lit" | "preact" | "react" | "svelte" | "vanilla" | "vue";

export type EntryOptions = {
  lib?: string,
  module?: boolean,
  global?: boolean,
}

export type PluginOptions = {
  include?: string[],
  exclude?: string[],
  config?: Config,
  configPath?: string,
  env?: Partial<Record<"config" | "utilities" | "variants", EntryOptions>> & {
    nodeModulesPath?: string,
    globalPath?: string,
    modulePath?: string,
  }
};

export type ResolvedPluginOptions = Required<PluginOptions>;

export type VitePlugin = {
  name: string;
  api: {
    vuePreprocess: (code: string, setup?: boolean) => string;
    sveltePreprocess: (ts?: { script: Function } & object) => ({
      script: (options: object) => Promise<{ code: string }>;
    })
  };
  resolveId(id: string): string | undefined;
  load(id: string): string | undefined;
  transform: (code: string, id: string) => string;
}
