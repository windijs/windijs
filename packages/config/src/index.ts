export * from "./bg";
export * from "./svg";
export * from "./flex";
export * from "./grid";
export * from "./border";
export * from "./effects";
export * from "./behaviors";
export * from "./filter";
export * from "./opacity";
export * from "./sizing";
export * from "./spacing";
export * from "./animation";
export * from "./transform";
export * from "./positioning";
export * from "./interactivity";
export * from "./transitions";
export * from "./typography";
export * from "./keyframes";
export * from "./interface";
