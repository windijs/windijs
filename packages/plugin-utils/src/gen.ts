import { Config, Handler, isStyleObject, mergeObject } from "@windijs/helpers";
import { indent, isNumber, isVarName } from "@windijs/shared";

import { injectHelper } from "./utils";

/**
 * Generate type interface for config object
 */
export function dtsConfig<T extends { DEFAULT?: unknown } & object>(
  obj: T,
  defaultType: string | ((v: unknown) => string) = "string",
  ignores: string[] = ["DEFAULT"],
  indentLevel = 0,
  indentation = 4
): string {
  const handleValue =
    typeof defaultType === "function"
      ? defaultType
      : (v: unknown) =>
          typeof v === "object" && v != null && !Array.isArray(v) ? dtsConfig(v, defaultType, ignores, indentLevel + 1, indentation) : defaultType;
  const contents = Object.entries(obj)
    .filter(([k]) => !ignores.includes(k))
    .map(([k, v]) => (isVarName(k) || isNumber(k) ? k : JSON.stringify(k)) + ": " + handleValue(v));
  return (
    ("DEFAULT" in obj ? handleValue(obj.DEFAULT) + " & " : "") +
    "{\n" +
    contents.map(i => indent(i, indentation * (indentLevel + 1))).join("\n") +
    "\n" +
    indent("}", indentation * indentLevel)
  );
}

/**
 * Generate type interface for setup object
 */
export function dtsSetup<T extends { DEFAULT?: unknown } & object>(
  obj: T,
  ignores: string[] = ["DEFAULT"],
  indentLevel = 0,
  indentation = 4
): string {
  const handleValue = (v: unknown) =>
    isStyleObject(v)
      ? "StyleObject"
      : v != null && typeof v === "object"
      ? "get" in v && "type" in v
        ? // eslint-disable-next-line @typescript-eslint/no-use-before-define
          dtsHandler(v as Handler<unknown>, ignores, indentLevel + 1, indentation)
        : dtsSetup(v, ignores, indentLevel + 1, indentation)
      : "undefined";
  return dtsConfig(obj, handleValue, ignores, indentLevel, indentation);
}

/**
 * Generate type interface for handler
 */
export function dtsHandler<R>(handler: Handler<R>, ignores: string[] = ["DEFAULT"], indentLevel = 0, indentation = 4): string {
  switch (handler.type) {
    case "config":
      return dtsConfig(handler.meta.config, "StyleObject", ignores, indentLevel, indentation);
    case "color":
      return dtsConfig(handler.meta.colors, handler.meta.op ? "ColorStyleObject" : "StyleObject", ignores, indentLevel, indentation);
    case "number":
      return "Record<0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12, StyleObject> & Record<number, StyleObject>";
    case "spacing":
      return "Record<0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 14 | 16 | 20 | 24 | 28 | 32 | 36 | 40 | 44 | 48 | 52 | 56 | 60 | 64 | 72 | 80 | 96, StyleObject> & Record<number, StyleObject>";
    case "fraction":
      return 'Record<"1/2" | "1/3" | "2/3" | "1/4" | "2/4" | "3/4" | "1/5" | "2/5" | "3/5" | "4/5" | "1/6" | "2/6" | "3/6" | "4/6" | "5/6", StyleObject> & { [key: string]: StyleObject }';
    case "guard":
      return `{ ${handler.meta.key}: ${dtsHandler(handler.meta.handler, ignores, indentLevel + 1, indentation)} }`;
    case "meld":
      return handler.meta.handlers.map(i => dtsHandler(i, ignores, indentLevel, indentation)).join(" & ");
    case "setup":
      return dtsSetup(handler.meta.config, ignores, indentLevel, indentation);
    default:
      return "{}";
  }
}

/**
 * Generate utilities.d.ts with config
 */
export function dtsUtilities(tmpl: string, config: Config): string {
  const theme = config.theme ?? {};
  let code = injectHelper(
    Object.entries(theme).reduce(
      (prev, [k, v]) =>
        v
          ? prev.replace(
              new RegExp(`"\\$windi\\.config\\.${k}Config.proxy"`, "g"),
              dtsConfig(theme.extend?.[k] ? mergeObject(v, theme.extend[k]!) : v, "StyleObject<{}>")
            )
          : prev,
      tmpl
    ),
    "MergeObject",
    "@windijs/helpers"
  );

  code = Object.entries(theme.extend ?? {})
    .filter(([k]) => !(k in theme))
    .reduce(
      (prev, [k, v]) =>
        v
          ? prev.replace(
              new RegExp(`Inject<(((?!Inject<)[\\s\\S])*),\\s*"\\$windi.config.${k}Config.proxy"`, "g"),
              `MergeObject<$1, ${dtsConfig(v, "StyleObject<{}>")}`
            )
          : prev,
      code
    );

  const colorStyle = "StyleObject<{opacity: (op: number) => StyleObject<{readonly gradient: StyleObject<{}>}>; readonly gradient: StyleObject<{}>}>";
  if (theme.colors) {
    const colors = theme.extend?.colors ? mergeObject(theme.colors, theme.extend.colors) : theme.colors;
    code = code
      .replace(/"\$windi\.config\.colorsConfig"/g, dtsConfig(colors))
      .replace(/"\$windi\.color\.colors\.proxy"/g, dtsConfig(colors, colorStyle));
  } else if (theme.extend?.colors)
    code = code
      .replace(/(const|var|let)(\s+colors:\s*)(Inject<[\s\S]*?);(?=\s*const)/, `$1$2MergeObject<$3, ${dtsConfig(theme.extend.colors)}>;`)
      .replace(/Inject<(((?!Inject<)[\s\S])*),\s*"\$windi.color.colors.proxy"/g, `MergeObject<$1, ${dtsConfig(theme.extend.colors, colorStyle)}`);

  return code;
}
