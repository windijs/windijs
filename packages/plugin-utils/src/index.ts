export * from "./plugin";
export * from "./types";
export * from "./utils";
export * from "./gen";
