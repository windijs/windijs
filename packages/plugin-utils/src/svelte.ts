import { StyleObject } from "@windijs/helpers";

export function w(node: HTMLElement, utilities: (StyleObject | StyleObject[])[]) {
  const apply = (a: (StyleObject | StyleObject[])[]) => a.flat().map(i => i.toString());
  let classes = apply(utilities);
  node.classList.add(...classes);
  return {
    update(utilities: (StyleObject | StyleObject[])[]) {
      node.classList.remove(...classes);
      classes = apply(utilities);
      node.classList.add(...classes);
    },

    destroy() {
      return;
    },
  };
}
