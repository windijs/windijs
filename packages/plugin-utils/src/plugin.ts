import { PluginOptions, SvelteVitePlugins, VitePlugins } from "./types";
import { configPath, declModule, filterConflict, injectConfig, injectHelper, injectImports, isProduction, readModule, replaceEntry, writeFile } from "./utils";
import { dirname, resolve } from "path";
import { dtsSetup, dtsUtilities } from "./gen";
import { existsSync, mkdirSync, readFileSync, readdirSync, rmSync, writeFileSync } from "fs";

import { Config } from "@windijs/config";

export const DefaultOptions: PluginOptions = {
  configPath: resolve("./windi.config"),
  env: {
    globalPath: "./src/windi-global.d.ts",
    modulePath: "./src/windi-module.d.ts",
    variants: {
      lib: "@windijs/variants",
      module: true,
      global: true,
    },
    utilities: {
      lib: "@windijs/utilities",
      module: true,
      global: true,
    },
    config: {
      module: true,
    },
  },
};

export function genVariants (options: PluginOptions) {
  const userVariants = options.config?.variants ?? {};
  const { mjs, dts } = readModule(options.env?.variants?.lib ?? "@windijs/variants");
  let code = injectHelper(injectConfig(replaceEntry(mjs)), "setupVariant");

  const defaults = dts.match(/[\w_$]+(?=:\s*VariantBuilder)/g) ?? [];
  const keys = Object.keys(userVariants);
  const items = defaults.concat(keys.filter(i => !defaults.includes(i)));

  const loaded = keys.filter(k => defaults.includes(k));
  const added = keys.filter(k => !defaults.includes(k));

  code = loaded.reduce((prev, k) => prev.replace(new RegExp(`[^;]*const\\s*${k}\\s*=[^;]*;`), ""), code);

  if (options.config?.variants) {
    code = code.replace(/(?=export\s\{[\s\S]*};?\s*$)/, `const { ${keys.join(", ")} } = setupVariant(windiUserConfig.variants);\n` + (added.length > 0 ? `export { ${added.join(", ")} };\n` : ""));
  }

  return {
    code,
    items,
  };
}

function injectTheme (code: string, config: Config) {
  const theme = config.theme ?? {};

  if (theme.colors) code = code.replace(/(const|let|var)\s+colors\s*=[\s\S]+?;/, "const colors = windiUserConfig.theme.colors;"); // replace colors

  for (const k in theme) {
    const regex = new RegExp(`(?<=configHandler\\().*?${k}Config`, "g");
    const matched = code.match(regex);
    if (matched) {
      code = code.replace(regex, "windiUserConfig.theme." + k);
    }
  }

  return code;
}

export function genProduction (options: PluginOptions) {
  const config = options.config ?? {};
  const variants = genVariants(options);

  let { pkg, mjs, dts } = readModule(options.env?.utilities?.lib ?? "@windijs/utilities");

  let code = mjs;

  const env = options.env ?? {};

  const defaults = code.match(/(?<=\s+as\s+)\S+/g) ?? [];

  // copy es folder and replace changed config
  pkg.exports["./__proxy__"] = {
    import: "./__proxy__/index.js",
  };
  writeFileSync(resolve("./node_modules/@windijs/utilities/package.json"), JSON.stringify(pkg, undefined, 2) + "\n");
  const esdir = resolve("./node_modules/@windijs/utilities", dirname(pkg.module));
  const pdir = "./node_modules/@windijs/utilities/__proxy__";
  if (existsSync(pdir)) rmSync(pdir, { recursive: true });
  mkdirSync(pdir);
  readdirSync(esdir).forEach(i => {
    const raw = readFileSync(resolve(esdir, i)).toString();
    writeFileSync(resolve(pdir, i), injectConfig(injectTheme(raw, config)));
  });

  const items = defaults;

  dts = replaceEntry(dtsUtilities(dts, config));
  let entry: string | undefined;

  for (const [k, v] of Object.entries(config.utilities ?? {})) {
    const t = `export declare const ${k}: ` + dtsSetup(v) + ";";
    const u = `const ${k} = setupUtility('${k}', windiUserConfig.utilities.${k})`;
    if (defaults.includes(k)) {
      // replace old utility type
      dts = dts.replace(new RegExp(`(export\\s+declare\\s+const|var|let)\\s${k}:[\\s\\S]*?;(?=(\\s*export\\s+declare\\s+(const|var|let)\\s[_$\\w])|(\\s*\\}\\s*$))`), t);
      entry = code.match(new RegExp(`(?<=export\\s*{[^}]*\\s+${k}\\s*}\\s*from\\s*['"])[^'"]+`))?.[0];
    } else {
      // add new utility type
      items.push(k);
      dts = dts + t + "\n";
      entry = undefined;
    }

    if (!entry) code += `export { default as ${k} } from './${k}.js';\n`;
    writeFileSync(resolve(pdir, entry ?? (k + ".js")), injectConfig(`import { setupUtility } from "@windijs/core";\n${u};\nexport { ${k} as default };\n`));
  }

  writeFileSync(resolve(pdir, "index.js"), code);

  const variantsGlobal = env.variants?.global ? (variants.items.map(i => `const ${i}: VariantBuilder;`).join("\n") + "\n") : undefined;
  let global = env.utilities?.global ? (dts.replace(/export\s+declare\s+/g, "").replace("type", "declare global {\n" + (variantsGlobal ?? "") + "type") + "}\n") : variantsGlobal ? ("declare global {\n" + variantsGlobal + "}\n") : undefined;
  if (global && env.variants?.global) global = injectImports(global, { windijs: ["VariantBuilder"] });

  const modules: string[] = [];
  if (env.config?.module) modules.push(declModule("virtual:config", ["export default (await import(\"../windi.config\")).default"]));
  if (env.variants?.module) modules.push(declModule("virtual:variants", ["import { VariantBuilder } from \"windijs\"", ...variants.items.map(i => `export declare const ${i}: VariantBuilder`)]));
  if (env.utilities?.module) modules.push(declModule("virtual:utilities", dts));

  return {
    global,
    module: modules.length > 0 ? modules.join("") : undefined,
    utilities: {
      code,
      items,
    },
    variants,
  };
}

export function genRuntime (options: PluginOptions) {
  const config = options.config ?? {};
  const variants = genVariants(options);

  let { mjs, dts } = readModule(options.env?.utilities?.lib ?? "@windijs/utilities", "dist/utilities.mjs"); // TODO: this entry is hard coded for now, maybe change to package exports later.

  const defaults = ["colors", ...(mjs.match(/(?<=createUtility\(")[\w_$-]+/g) ?? [])];
  const env = options.env ?? {};
  const items = defaults;

  let code = injectTheme(injectHelper(injectConfig(replaceEntry(mjs)), "setupUtility"), config);
  dts = replaceEntry(dtsUtilities(dts, config));

  for (const [k, v] of Object.entries(config.utilities ?? {})) {
    const t = `export declare const ${k}: ` + dtsSetup(v) + ";";
    const u = `const ${k} = setupUtility('${k}', windiUserConfig.utilities.${k})`;
    if (defaults.includes(k)) {
      // replace old utility type
      dts = dts.replace(new RegExp(`(export\\s+declare\\s+const|var|let)\\s${k}:[\\s\\S]*?;(?=(\\s*export\\s+declare\\s+(const|var|let)\\s[_$\\w])|(\\s*\\}\\s*$))`), t);
      code = code.replace(new RegExp(`(const|let|var)\\s*${k}\\s*=[^;]*`), u);
    } else {
      // add new utility type
      items.push(k);
      dts = dts + t + "\n";
      code = code.replace(/(?=export\s\{[\s\S]*};?\s*$)/, "export " + u + ";\n");
    }
  }

  const variantsGlobal = env.variants?.global ? (variants.items.map(i => `const ${i}: VariantBuilder;`).join("\n") + "\n") : undefined;
  let global = env.utilities?.global ? (dts.replace(/export\s+declare\s+/g, "").replace("type", "declare global {\n" + (variantsGlobal ?? "") + "type") + "}\n") : variantsGlobal ? ("declare global {\n" + variantsGlobal + "}\n") : undefined;
  if (global && env.variants?.global) global = injectImports(global, { windijs: ["VariantBuilder"] });

  const modules: string[] = [];
  if (env.config?.module) modules.push(declModule("virtual:config", ["export default (await import(\"../windi.config\")).default"]));
  if (env.variants?.module) modules.push(declModule("virtual:variants", ["import { VariantBuilder } from \"windijs\"", ...variants.items.map(i => `export declare const ${i}: VariantBuilder`)]));
  if (env.utilities?.module) modules.push(declModule("virtual:utilities", dts));

  return {
    global,
    module: modules.length > 0 ? modules.join("") : undefined,
    utilities: {
      code,
      items,
    },
    variants,
  };
}

export function virtualUtilities (utilities: string) {
  const virtualModuleId = "virtual:utilities";
  const resolvedVirtualModuleId = "\0" + virtualModuleId;

  return {
    name: "virtual-utilities",
    resolveId (id: string) {
      if (id === virtualModuleId) return resolvedVirtualModuleId;
    },
    load (id: string) {
      if (id === resolvedVirtualModuleId) return utilities;
    },
  };
}

export function virtualVariants (variants: string) {
  const virtualModuleId = "virtual:variants";
  const resolvedVirtualModuleId = "\0" + virtualModuleId;

  return {
    name: "virtual-variants",
    resolveId (id: string) {
      if (id === virtualModuleId) return resolvedVirtualModuleId;
    },
    load (id: string) {
      if (id === resolvedVirtualModuleId) return variants;
    },
  };
}

export function virtualConfig () {
  const virtualModuleId = "virtual:config";

  return {
    name: "virtual-config",
    resolveId (id: string) {
      if (id === virtualModuleId) return configPath;
    },
  };
}

export function createRuntime (options: PluginOptions = {}) {
  options = Object.assign(DefaultOptions, options);
  const { global, module, utilities, variants } = isProduction ? genProduction(options) : genRuntime(options);

  writeFile(options.env?.globalPath ?? "./src/windi-global.d.ts", global);
  writeFile(options.env?.modulePath ?? "./src/windi-module.d.ts", module);

  function preprocess (code: string) {
  // TODO: fix: The $ prefix is reserved, and cannot be used for variable and import names
    return injectImports(code, {
      [isProduction ? "@windijs/utilities/__proxy__" : "virtual:utilities"]: filterConflict(code, utilities.items),
      "virtual:variants": filterConflict(code, variants.items).filter(i => !i.startsWith("$")),
    });
  }

  return { utilities: utilities.code, variants: variants.code, preprocess };
}

export const vitePlugins = new Proxy({}, {
  get (target, p) {
    return (options: PluginOptions = {}) => {
      const { utilities, variants, preprocess } = createRuntime(options);
      const basePlugins = [
        virtualConfig(),
        virtualUtilities(utilities),
        virtualVariants(variants),
      ];

      function vanillaPreprocess (exts = [".js", ".ts"]) {
        return {
          name: "vanilla-preprocess-windijs",
          transform (code: string, id: string) {
            if (exts.find(i => id.endsWith(i))) {
              return preprocess(code);
            }
            return code;
          },
        };
      }

      if (p === "vanilla") {
        return {
          plugins: basePlugins,
          windijs: vanillaPreprocess,
        };
      }

      if (p === "vue") {
        function vuePreprocess (exts = [".vue"]) {
          return {
            name: "vue-preprocess-windijs",
            transform (code: string, id: string) {
              if (id.endsWith(".vue")) {
                const regex = /<script.*>/;
                const match = regex.exec(code);
                if (!match) { return `<script>\n${preprocess("")}</script>` + code; }
                const start = regex.lastIndex + match[0].length;
                return code.slice(0, start) + "\n" + preprocess(code.slice(start));
              }
              return vanillaPreprocess(exts);
            },
          };
        }

        return {
          plugins: basePlugins,
          windijs: vuePreprocess,
        };
      };

      if (p === "svelte") {
        function sveltePreprocess (ts: { script: Function } & object) {
          return {
            script: async (options: object) => {
              // @ts-ignore
              if (options.attributes?.lang !== "ts") return { code: preprocess(options.content) };
              const res = await ts.script(options);
              res.code = preprocess(res.code);
              return res;
            },
          };
        }

        return {
          plugins: basePlugins,
          windijs: sveltePreprocess,
        };
      }
    };
  },
}) as {
  vue: (options: PluginOptions) => VitePlugins,
  vanilla: (options: PluginOptions) => VitePlugins,
  svelte: (options: PluginOptions) => SvelteVitePlugins
};
