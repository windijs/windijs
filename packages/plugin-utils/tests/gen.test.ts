import { backgroundSizeConfig, borderRadiusConfig, touchActionConfig } from "@windijs/config";
import { css } from "@windijs/helpers";

import { dtsConfig, dtsHandler, dtsUtilities } from "../src/gen";

const get = (prop: string) => prop;

test("dts", () => {
  expect(dtsConfig(backgroundSizeConfig)).toMatchSnapshot();
});

test("dts nested", () => {
  expect(dtsConfig(touchActionConfig)).toMatchSnapshot();
  expect(
    dtsConfig({
      a: {
        b: {
          c: {
            d: "d",
          },
        },
      },
    })
  ).toMatchSnapshot();
});

test("dts number and string", () => {
  expect(
    dtsConfig({
      $: "a",
      2: "b",
      1.5: "c",
      "dts-test": "d",
    })
  ).toMatchSnapshot();
});

test("dts with default type", () => {
  expect(dtsConfig(touchActionConfig, "StyleObject<{}>")).toMatchSnapshot();
});

test("dts ignore DEFAULT", () => {
  expect(dtsConfig(borderRadiusConfig, "StyleObject<{}>", ["DEFAULT"])).toMatchSnapshot();
});

test("dst with array should be replaced", () => {
  expect(dtsConfig({ a: ["a", "b"] })).toMatchSnapshot();
});

test("generate dts", () => {
  const tmpl = `import { buildLinearGradient, buildTransition } from "@windijs/core";
export declare const image: {
  render: Inject<{
      auto: StyleObject<{}>;
      pixel: StyleObject<{}>;
      edge: StyleObject<{}>;
  }, "$windi.config.imageRenderingConfig.proxy">;
};
export declare const fill: Inject<{}, "$windi.color.colors.proxy">;
export declare const colors: Inject<{}, "$windi.config.colorsConfig">;
`;

  expect(
    dtsUtilities(tmpl, {
      theme: {
        colors: {
          blue: {
            50: "#ecfeff",
            100: "#cffafe",
          },
          cyan: {
            50: "#ecfeff",
            100: "#cffafe",
            200: "#a5f3fc",
          },
        },
        imageRendering: {
          auto: "auto",
          pixel: "pixelated",
          edge: "crisp-edges",
        },
      },
    })
  ).toMatchSnapshot();
});

test("dtsHandler with configHandler", () => {
  expect(
    dtsHandler({
      type: "config",
      meta: {
        config: {
          fixed: "fixed",
          local: "local",
          scroll: "scroll",
        },
      },
      get,
    })
  ).toMatchSnapshot();

  expect(
    dtsHandler({
      type: "config",
      meta: {
        config: {
          DEFAULT: "repeat",
          x: "repeat-x",
          y: "repeat-y",
          round: "round",
          space: "space",
        },
      },
      get,
    })
  ).toMatchSnapshot();

  expect(
    dtsHandler({
      type: "config",
      meta: {
        config: {
          repeat: {
            DEFAULT: "repeat",
            x: "repeat-x",
            y: "repeat-y",
            round: "round",
            space: "space",
          },
          noRepeat: "no-repeat",
        },
      },
      get,
    })
  ).toMatchSnapshot();
});

test("dtsHandler with colorHandler", () => {
  expect(
    dtsHandler({
      type: "color",
      meta: {
        colors: {
          red: "500",
        },
        op: undefined,
      },
      get,
    })
  ).toMatchSnapshot();

  expect(
    dtsHandler({
      type: "color",
      meta: {
        colors: {
          red: "500",
        },
        op: "--windi-opacity",
      },
      get,
    })
  ).toMatchSnapshot();
});

test("dtsHandler with numberHandler", () => {
  expect(
    dtsHandler({
      type: "number",
      meta: {
        size: "",
      },
      get,
    })
  ).toMatchSnapshot();

  expect(
    dtsHandler({
      type: "number",
      meta: {
        size: "rem",
      },
      get,
    })
  ).toMatchSnapshot();

  expect(
    dtsHandler({
      type: "fraction",
      meta: undefined,
      get,
    })
  ).toMatchSnapshot();
});

test("dtsHandler with guard", () => {
  expect(
    dtsHandler({
      type: "guard",
      meta: {
        key: "size",
        handler: {
          type: "number",
          meta: {
            size: "",
          },
          get,
        },
      },
      get,
    })
  ).toMatchSnapshot();
});

test("dtsHandler with meld", () => {
  expect(
    dtsHandler({
      type: "meld",
      meta: {
        handlers: [
          {
            type: "config",
            meta: {
              config: {
                DEFAULT: "repeat",
                x: "repeat-x",
                y: "repeat-y",
                round: "round",
                space: "space",
              },
            },
            get,
          },
          {
            type: "guard",
            meta: {
              key: "size",
              handler: {
                type: "number",
                meta: {
                  size: "",
                },
                get,
              },
            },
            get,
          },
        ],
      },
      get,
    })
  ).toMatchSnapshot();
});

test("dtsHandler with setup", () => {
  expect(
    dtsHandler({
      type: "setup",
      meta: {
        config: {
          red: css({ backgroundColor: "red" }),
          blue: {
            light: css({ backgroundColor: "blueviolet" }),
            handle: {
              type: "config",
              meta: {
                config: {
                  DEFAULT: "repeat",
                  x: "repeat-x",
                  y: "repeat-y",
                  round: "round",
                  space: "space",
                },
              },
              get,
            },
          },
        },
      },
      get,
    })
  ).toMatchSnapshot();

  expect(
    dtsHandler({
      type: "setup",
      meta: {
        config: {
          red: css({ backgroundColor: "red" }),
          blue: {
            light: css({ backgroundColor: "blueviolet" }),
            handle: {
              type: "meld",
              meta: {
                handlers: [
                  {
                    type: "config",
                    meta: {
                      config: {
                        DEFAULT: "repeat",
                        x: "repeat-x",
                        y: "repeat-y",
                        round: "round",
                        space: "space",
                      },
                    },
                    get,
                  },
                  {
                    type: "guard",
                    meta: {
                      key: "size",
                      handler: {
                        type: "number",
                        meta: {
                          size: "",
                        },
                        get,
                      },
                    },
                    get,
                  },
                ],
              },
              get,
            },
          },
        },
      },
      get,
    })
  ).toMatchSnapshot();
});
