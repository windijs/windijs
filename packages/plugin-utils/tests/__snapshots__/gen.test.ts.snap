// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dst with array should be replaced 1`] = `
"{
    a: string
}"
`;

exports[`dts 1`] = `
"{
    auto: string
    cover: string
    contain: string
}"
`;

exports[`dts ignore DEFAULT 1`] = `
"StyleObject<{}> & {
    1: StyleObject<{}>
    none: StyleObject<{}>
    sm: StyleObject<{}>
    md: StyleObject<{}>
    lg: StyleObject<{}>
    xl: StyleObject<{}>
    xxl: StyleObject<{}>
    xxxl: StyleObject<{}>
    full: StyleObject<{}>
}"
`;

exports[`dts nested 1`] = `
"{
    auto: string
    none: string
    pan: {
        x: string
        y: string
        left: string
        right: string
        up: string
        down: string
    }
    \\"pinch-zoom\\": string
    manipulation: string
}"
`;

exports[`dts nested 2`] = `
"{
    a: {
        b: {
            c: {
                d: string
            }
        }
    }
}"
`;

exports[`dts number and string 1`] = `
"{
    2: string
    $: string
    1.5: string
    \\"dts-test\\": string
}"
`;

exports[`dts with default type 1`] = `
"{
    auto: StyleObject<{}>
    none: StyleObject<{}>
    pan: {
        x: StyleObject<{}>
        y: StyleObject<{}>
        left: StyleObject<{}>
        right: StyleObject<{}>
        up: StyleObject<{}>
        down: StyleObject<{}>
    }
    \\"pinch-zoom\\": StyleObject<{}>
    manipulation: StyleObject<{}>
}"
`;

exports[`dtsHandler with colorHandler 1`] = `
"{
    red: StyleObject
}"
`;

exports[`dtsHandler with colorHandler 2`] = `
"{
    red: ColorStyleObject
}"
`;

exports[`dtsHandler with configHandler 1`] = `
"{
    fixed: StyleObject
    local: StyleObject
    scroll: StyleObject
}"
`;

exports[`dtsHandler with configHandler 2`] = `
"StyleObject & {
    x: StyleObject
    y: StyleObject
    round: StyleObject
    space: StyleObject
}"
`;

exports[`dtsHandler with configHandler 3`] = `
"{
    repeat: StyleObject & {
        x: StyleObject
        y: StyleObject
        round: StyleObject
        space: StyleObject
    }
    noRepeat: StyleObject
}"
`;

exports[`dtsHandler with guard 1`] = `"{ size: Record<0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12, StyleObject> & Record<number, StyleObject> }"`;

exports[`dtsHandler with meld 1`] = `
"StyleObject & {
    x: StyleObject
    y: StyleObject
    round: StyleObject
    space: StyleObject
} & { size: Record<0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12, StyleObject> & Record<number, StyleObject> }"
`;

exports[`dtsHandler with numberHandler 1`] = `"Record<0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12, StyleObject> & Record<number, StyleObject>"`;

exports[`dtsHandler with numberHandler 2`] = `"Record<0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12, StyleObject> & Record<number, StyleObject>"`;

exports[`dtsHandler with numberHandler 3`] = `"Record<\\"1/2\\" | \\"1/3\\" | \\"2/3\\" | \\"1/4\\" | \\"2/4\\" | \\"3/4\\" | \\"1/5\\" | \\"2/5\\" | \\"3/5\\" | \\"4/5\\" | \\"1/6\\" | \\"2/6\\" | \\"3/6\\" | \\"4/6\\" | \\"5/6\\", StyleObject> & { [key: string]: StyleObject }"`;

exports[`dtsHandler with setup 1`] = `
"{
    red: StyleObject
    blue: {
        light: StyleObject
        handle: StyleObject & {
            x: StyleObject
            y: StyleObject
            round: StyleObject
            space: StyleObject
        }
    }
}"
`;

exports[`dtsHandler with setup 2`] = `
"{
    red: StyleObject
    blue: {
        light: StyleObject
        handle: StyleObject & {
            x: StyleObject
            y: StyleObject
            round: StyleObject
            space: StyleObject
        } & { size: Record<0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12, StyleObject> & Record<number, StyleObject> }
    }
}"
`;

exports[`generate dts 1`] = `
"import { buildLinearGradient, buildTransition } from \\"@windijs/core\\";
export declare const image: {
  render: Inject<{
      auto: StyleObject<{}>;
      pixel: StyleObject<{}>;
      edge: StyleObject<{}>;
  }, {
    auto: StyleObject<{}>
    pixel: StyleObject<{}>
    edge: StyleObject<{}>
}>;
};
export declare const fill: Inject<{}, {
    blue: {
        50: StyleObject<{opacity: (op: number) => StyleObject<{readonly gradient: StyleObject<{}>}>; readonly gradient: StyleObject<{}>}>
        100: StyleObject<{opacity: (op: number) => StyleObject<{readonly gradient: StyleObject<{}>}>; readonly gradient: StyleObject<{}>}>
    }
    cyan: {
        50: StyleObject<{opacity: (op: number) => StyleObject<{readonly gradient: StyleObject<{}>}>; readonly gradient: StyleObject<{}>}>
        100: StyleObject<{opacity: (op: number) => StyleObject<{readonly gradient: StyleObject<{}>}>; readonly gradient: StyleObject<{}>}>
        200: StyleObject<{opacity: (op: number) => StyleObject<{readonly gradient: StyleObject<{}>}>; readonly gradient: StyleObject<{}>}>
    }
}>;
export declare const colors: Inject<{}, {
    blue: {
        50: string
        100: string
    }
    cyan: {
        50: string
        100: string
        200: string
    }
}>;
"
`;
